runJupyterDaskOnSLURM.py

```diff
@@ -0,0 +1,268 @@
+    retrieval functions.
+
+    :param args: ArgumentParser return object containing command line arguments
+
+    :return config_inputs: dictionary with fields {host: user: keypath:}
+    :return platform: (echo) platform name
+    """
+    if args.add_platform:
+	    return add_platform()
+    elif args.one_off:
+	    return add_platform(oneoff=True)
+    else:
+        return load_platform_config(args.platform)
+
+def establish_connection(config_inputs):
+    """
+    Establish ssh connection oject to remote host based on configuration.
+
+    :param config_inputs: configuration input parameters
+    :return conn: ssh connection object
+    """
+    conn = Connection(host=config_inputs['host'],user=config_inputs['user'],connect_kwargs={'key_filename':config_inputs['keypath']})
+    return conn
+
+def submit_scheduler(conn, args, platform):
+    """
+    submit batch job with jupyter server on remote host
+
+    :param conn: ssh connection object
+    :param args: ArgumentParser return object containing command line arguments. included for optional specification of local port
+    :return outfilename: name of slurm output file on remote host
+    """
+    jscript = 'jupyter_dask_'+platform+'.bsh'
+    if args.local_port is not None:
+        lport = args.local_port
+        cmd = f"cd {remoteWD} && sbatch --export=ALL,lport={lport} {remoteScriptD}{jscript}"
+    else:
+        cmd = f"cd {remoteWD} && sbatch {remoteScriptD}{jscript}"
+    result = conn.run(cmd)
+    jobnumber = result.stdout.rstrip().split(' ')[3]
+    outfilename = 'slurm-'+jobnumber+'.out'
+    return outfilename
+
+def retrieve_node_info(conn,outfilename):
+    """
+    Parse and retrieve information on node where server is running on remote platform and whicch ports are being used
+    for port forwarding to local host
+
+    :param conn: ssh connection object
+    :param outfilename: name of slurm output file on remote host
+    :return portsnodes: pastable string for ssh port forwarding command
+    :return lp: local port number
+    :return node: node identifier of remote platform
+    :return remoteport: number of remote port where jupyter server is exposed
+    """
+    cmd = f"cd {remoteWD} && cat {outfilename} | grep '/path/to/private/ssh/key' - | cut -d ' ' -f 6- - "
+    result = conn.run(cmd)
+    portsnodes = result.stdout.rstrip()
+    node = portsnodes.split(' ')[0].split(':')[1]
+    lp = portsnodes.split(' ')[0].split(':')[0]
+    remoteport = portsnodes.split(' ')[0].split(':')[2] 
+    return portsnodes, lp, node, remoteport
+
+def forward_ports(portsnodes, config_inputs):
+    """
+    execute port forwarding command
+
+    :param portsnodes: pastable string for ssh port forwarding command
+    :config_inputs: dictionary with fields {host: user: keypath:} for ssh connection to remote 
+    """
+    cmd = f"ssh -i {config_inputs['keypath']} -N -L {portsnodes} & "
```

Comment:
By running the ssh command here with `&` the scripts open the browser window and terminates, leaving the ssh forwarding running in the background on the local system. 
What do you think about a structure like the following, using another fabric's Connection and its [forward_local](https://docs.fabfile.org/en/2.5/api/connection.html#fabric.connection.Connection.forward_local) method:

```python
with Connection(...).forward_local(...):
  launchJupyterLabLocal(localport)
  some_function_that_sleeps_indefinitely()
```
So the script would open the browser and hang, but the port forwarding would be stop when the program is killed. Potential improvement could also be to check in the slurm job whether the Jupyter server has started, and open the browser only then.

Reply:
I've replaced the port forwarding as suggested. The script waits for user input to finish, allowing the user to terminate the port forwarding.